tests/integration/test_rerelate.py

```diff
@@ -40,82 +45,111 @@ async def deploy(self, ops_test: OpsTest):
     )
-    await ops_test.model.wait_for_idle(status="active", timeout=1000)
+    await ops_test.model.wait_for_idle(status="active", timeout=600)
     assert await is_loki_up(ops_test, app_name)
 
 
 @pytest.mark.abort_on_fail
 async def test_remove_related_app(ops_test: OpsTest):
     await asyncio.gather(
-        *[ops_test.model.applications[app.name].remove() for app in related_apps],
-        # Block until it is really gone. Added after an itest failed when tried to redeploy:
-        # juju.errors.JujuError: ['cannot add application "...": application already exists']
-        *[
-            ops_test.model.block_until(lambda: app.name not in ops_test.model.applications)
-            for app in related_apps
-        ],
+        ops_test.model.applications["loki-tester"].remove(),
+        ops_test.model.applications["alertmanager"].remove(),
     )
-
-    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=1000)
+    logger.debug("Applications removed. Blocking for 60 seconds then force removing...")
+    # Block until it is really gone. Added after an itest failed when tried to redeploy:
+    # juju.errors.JujuError: ['cannot add application "...": application already exists']
+    try:
+        await ops_test.model.block_until(
+            lambda: "loki-tester" not in ops_test.model.applications,
+            lambda: "alertmanager" not in ops_test.model.applications,
+            timeout=60,
+        )
+    except asyncio.exceptions.TimeoutError:
+        logger.warning("Timeout reached while blocking!")
+
+    for app in filter(lambda x: x in ops_test.model.applications, ["loki-tester", "alertmanager"]):
+        cmd = [
+            "juju",
+            "remove-application",
+            "--destroy-storage",
+            "--force",
+            "--no-wait",
+            app,
+        ]
+        logger.info("Forcibly removing {}".format(app))
+        await ops_test.run(*cmd)
+
+    await ops_test.model.wait_for_idle(apps=[app_name], timeout=600)
     assert await is_loki_up(ops_test, app_name)
 
 
 @pytest.mark.abort_on_fail
-async def test_rerelate_app(ops_test: OpsTest):
+async def test_rerelate_app(ops_test: OpsTest, loki_tester_charm):
     await asyncio.gather(
-        *[app.deploy(ops_test) for app in related_apps],
+        ops_test.model.deploy(
+            loki_tester_charm,
+            resources=tester_resources,
+            application_name="loki-tester-new",
+        ),
+        ops_test.model.deploy(
+            "ch:alertmanager-k8s",
+            application_name="alertmanager-new",
+            channel="edge",
+        ),
     )
+
     await asyncio.gather(
-        *[ops_test.model.add_relation(app_name, app.name) for app in related_apps],
+        ops_test.model.add_relation(app_name, "loki-tester-new"),
+        ops_test.model.add_relation(app_name, "alertmanager-new"),
     )
+    await ops_test.model.wait_for_idle(status="active", timeout=600)
```

Comment:
Without the `apps` arg, `wait_for_idle` will complain (and eventually timeout) that `Waiting for model: ... (not enough units yet - 0/1)`, even though the unit that is `0/1` is the unit that was removed in the previous test function.

```suggestion
    await ops_test.model.wait_for_idle(apps=[app_name], status="active", timeout=600)
```

This change would hide the fact that some other app wasn't entirely removed.

Haven't checked yet what happens without the `-new`.

Reply:
I don't understand what you're suggesting here.

Why would we _want_ to add an argument which could possibly mask undesirable behavior?

Reply:
There is a double overlook here, which unmasked the error:
1. Calling wait_for_idle without appropriate args would hang as long as there is any app with zero units - technically, even if to speed things up, we typically specify the apps of interest.
2. Calling app.remove doesn't guarantee removal. That prompts for another assertion with block_until after remove, i.e. the fact that a previous test func left dirty state is not a concern for the next test function (this one).

Reply:
- There's only one thing to fail on here: when after a `remove`, our `block_until "app" not in applications" times out. https://github.com/juju/python-libjuju/issues/656
- But since we did not have that `block_until`, the next test function starts with leftovers.
- And because we didn't specify `apps=[...]` in `wait_for_idle`, you'd keep seeing `not enough units yet - 0/1` for the leftover unit -- it is a feature of `wait_for_idle` to keep waiting if the number of units is still zero:
  - https://github.com/juju/python-libjuju/issues/574
  - https://github.com/juju/python-libjuju/issues/608

Reply:
I really don't understand what the context is here anymore.

From my POV, there is only one thing to fail on:

* The state of the model does not match what our tests say it should. That's it.

This is basic test idempotency. If the code says "Install `foo`, `bar`, and `baz`, then remove `foo` and `baz` and wait for `bar` to be ready", I don't want to know _only_ if `bar` is ready, I want to know that the state of the environment matches what I expect, which _also_ includes the things I told to go away actually going away.

Seeing `not enough units yet - 0/1` for a "leftover" unit which we told to go away and did not and having the test fail is **exactly** the point of integration testing from my perspective. We talk a lot about testing the "happy path" and not having tests for unknown bugs. Well, we do, accidentally, and we've been working at it for most of the last week. This is not a bad or undesirable thing.

Reply:
Sorry, my `apps=[app_name]` must have caused the confusion: this was just an example - of course you list there all the apps you expect to see: `"loki-tester-new"` and `"alertmanager-new"`.
However, the `not enough units yet - 0/1` is for `"loki-tester"` or `"alertmanager"`. What I'm saying is that "The state of the model does not match what our tests say it should" for `"loki-tester"` or `"alertmanager"` should be covered by the previous test - `test_remove_related_app`, not this one.